packages/@aws-cdk/aws-ec2/lib/instance.ts

```diff
@@ -384,16 +384,27 @@ export class Instance extends Resource implements IInstance {
     this.applyUpdatePolicies(props);
 
     // Trigger replacement (via new logical ID) on user data change, if specified or cfn-init is being used.
+    //
+    // This is slightly tricky -- we need to resolve the UserData string (in order to get at actual Asset hashes,
+    // instead of the Token stringifications of them ('${Token[1234]}'). However, in the case of CFN Init usage,
+    // a UserData is going to contain the logicalID of the resource itself, which means infinite recursion if we
+    // try to naively resolve. We need a recursion breaker in this.
     const originalLogicalId = Stack.of(this).getLogicalId(this.instance);
+    let recursing = false;
     this.instance.overrideLogicalId(Lazy.uncachedString({
-      produce: () => {
-        let logicalId = originalLogicalId;
-        if (props.userDataCausesReplacement ?? props.initOptions) {
-          const md5 = crypto.createHash('md5');
-          md5.update(this.userData.render());
-          logicalId += md5.digest('hex').substr(0, 16);
+      produce: (context) => {
+        if (recursing) { return originalLogicalId; }
+        if (!(props.userDataCausesReplacement ?? props.initOptions)) { return originalLogicalId; }
+
+        const md5 = crypto.createHash('md5');
+        recursing = true;
+        try {
+          md5.update(JSON.stringify(context.resolve(this.userData.render())));
+        } finally {
+          recursing = false;
```

Comment:
For my own edification, is this just paranoia, or is there a legitimate case where not resetting this would cause issues? Typically, any errors thrown are going to stop synthesis in its tracks, right? Is this guarding against a case of someone writing their own custom synthesizer, or is there something in the "native" CDK that merits this `finally`?

Reply:
I think it might just be paranoia :)